nifi-extension-bundles/nifi-standard-services/nifi-oauth2-provider-bundle/nifi-oauth2-provider-service/src/main/java/org/apache/nifi/oauth2/JWTBearerOAuth2AccessTokenProvider.java

```diff
@@ -0,0 +1,677 @@
+                            final byte[] bytes = messageDigest.digest(certificates[0].getEncoded());
+                            String url = Base64.getEncoder().encodeToString(bytes);
+                            headerBuilder.x509CertSHA256Thumbprint(new Base64URL(url));
+                        }
+                    }
+                }
+            } catch (Exception e) {
+                getLogger().warn("Failed to set x5t header, setting to null.", e);
+            }
+        }
+
+        Map<String, String> formParams = new HashMap<>();
+        formParams.put("grant_type", grantType);
+        formParams.put(assertion, getAssertion(headerBuilder.build(), claimsSetBuilder.build()));
+        formParams.putAll(this.formParams);
+
+        requestTokenEndpoint(formParams);
+    }
+
+    protected void requestTokenEndpoint(Map<String, String> formParams) throws URISyntaxException {
+        HttpRequestHeadersSpec request = webClientService
+                .post()
+                .uri(new URI(tokenEndpoint))
+                .header("Accept", ContentType.APPLICATION_JSON.getType())
+                .header("Content-Type", ContentType.APPLICATION_URLENCODED.getType())
+                .body(formParams.entrySet()
+                        .stream()
+                        .map(param -> param.getKey() + "=" + param.getValue())
+                        .collect(Collectors.joining("&")));
+
+        try (final HttpResponseEntity response = request.retrieve()) {
+            if (response.statusCode() != 200) {
+                final String message = "Failed to retrieve access details. Status code: %s, response body: %s".formatted(
+                        response.statusCode(),
+                        readBodyAsText(response).orElse("[no body]"));
+
+                throw new AccessTokenRetrievalException(message);
+            }
+
+            try (final InputStream body = response.body()) {
+                accessDetails = ACCESS_DETAILS_MAPPER.readValue(body, AccessToken.class);
+            }
+        } catch (final IOException e) {
+            throw new AccessTokenRetrievalException("Failed to retrieve or process access token details", e);
+        }
+    }
+
+    protected String getAssertion(JWSHeader jwsHeader, JWTClaimsSet jwtClaimsSet) throws JOSEException {
+        SignedJWT signedJWT = new SignedJWT(jwsHeader, jwtClaimsSet);
+        signedJWT.sign(signer);
+        return signedJWT.serialize();
+    }
+
+    private Optional<String> readBodyAsText(final HttpResponseEntity response) {
+        try (final InputStream is = response.body()) {
+            final String body = new String(is.readAllBytes(), StandardCharsets.UTF_8);
+            return Optional.of(body);
+        } catch (final IOException e) {
+            getLogger().warn("Failed to read response body", e);
+            return Optional.empty();
+        }
+    }
+
+    private void getProperties(ConfigurationContext context) {
```

Comment:
This seems like it should be named `setProperties()` since it doesn't return anything.